Content.Server/Radio/EntitySystems/HeadsetSystem.cs

```diff
@@ -78,29 +100,116 @@ public void SetEnabled(EntityUid uid, bool value, HeadsetComponent? component =
+                    ("freq", proto.Frequency)));
+            }
+            args.PushMarkup(Loc.GetString("examine-headset-chat-prefix", ("prefix", ";")));
+        }
+    }
+
+    private void OnStartup(EntityUid uid, HeadsetComponent component, ComponentStartup args)
+    {
+        component.KeyContainer = _container.EnsureContainer<Container>(uid, HeadsetComponent.KeyContainerName);
+    }
+
+    private bool InstallKey(HeadsetComponent src, EntityUid key)
+    {
+        if(src.KeyContainer.Insert(key))
+        {
+            ++src.KeysInstalledAmount;
+            return true;
+        }
+        return false;
+    }
 
-        args.PushMarkup(Loc.GetString("examine-headset"));
+    private void RecalculateChannels(HeadsetComponent src)
+    {
+        src.Channels.Clear();
+        src.KeysInstalledAmount = 0;
+        foreach(EntityUid i in src.KeyContainer.ContainedEntities)
+            if(TryComp<EncryptionKeyComponent?>(i, out var key))
+            {
+                ++src.KeysInstalledAmount;
+                foreach(var j in key.Channels)
+                    src.Channels.Add(j);
+            }
+        return;
+    }
 
-        foreach (var id in component.Channels)
+    private void OnInteractUsing(EntityUid uid, HeadsetComponent component, InteractUsingEvent args)
+    {
+        if(!component.IsKeysExtractable || !TryComp<ContainerManagerComponent>(uid, out var Storage))
         {
-            if (id == "Common") continue;
-
-            var proto = _protoManager.Index<RadioChannelPrototype>(id);
-            args.PushMarkup(Loc.GetString("examine-headset-channel",
-                ("color", proto.Color),
-                ("key", proto.KeyCode),
-                ("id", proto.LocalizedName),
-                ("freq", proto.Frequency)));
+            return;
         }
+        if(TryComp<EncryptionKeyComponent?>(args.Used, out var key))
+        {
+            if(component.KeySlotsAmount > component.KeysInstalledAmount)
+                if(_container.TryRemoveFromContainer(args.Used) && InstallKey(component, args.Used))
+                {
+                    _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-successfully-installed"), uid, Filter.Entities(args.User));
+                    SoundSystem.Play(component.KeyInsertionSound.GetSound(), Filter.Pvs(args.Target), args.Target);
+                }
+            else
+                _popupSystem.PopupEntity(Loc.GetString("headset-encryption-key-slots-already-full"), uid, Filter.Entities(args.User));
+            return;
+        } 
+        if(TryComp<ToolComponent?>(args.Used, out var tool))
+        {
+            if(component.KeysInstalledAmount > 0)
+            {
+                if(_toolSystem.UseTool(
+                    args.Used, args.User, uid,
+                    0f, 0f, new String[] { "Screwing" },
+                    doAfterCompleteEvent: null, toolComponent: tool)
+                )
+                {
```

Comment:
You're using almost nothing of usetool so why even call it here? also doaftercompleteevent is nullable.

Reply:
To check is it tool with Screwing.
Do after complete nulled cause UseTool messing with so many its overloads.